crates/biome_yaml_parser/src/lexer/mod.rs

```diff
@@ -233,3 +294,103 @@ impl<'src> Lexer<'src> for YamlLexer<'src> {
         }
     }
 }
+
+impl<'src> ReLexer<'src> for YamlLexer<'src> {
+    fn re_lex(&mut self, context: Self::ReLexContext) -> Self::Kind {
+        self.position = u32::from(self.current_start) as usize;
+
+        let kind = match self.current_byte() {
+            Some(current) => self.consume_token(current, context),
+            None => EOF,
+        };
+        self.current_kind = kind;
+        kind
+    }
+}
+
+impl<'src> LexerWithCheckpoint<'src> for YamlLexer<'src> {
+    fn checkpoint(&self) -> LexerCheckpoint<Self::Kind> {
+        LexerCheckpoint {
+            position: TextSize::from(self.position as u32),
+            current_start: self.current_start,
+            current_flags: self.current_flags,
+            current_kind: self.current_kind,
+            after_line_break: self.has_preceding_line_break(),
+            unicode_bom_length: self.unicode_bom_length,
+            diagnostics_pos: self.diagnostics.len() as u32,
+        }
+    }
+}
+
+/// Lex context as specified in
+/// https://yaml.org/spec/1.2.2/#42-production-parameters
+#[derive(Default, Clone, Copy)]
+pub enum YamlLexContext {
+    #[default]
+    Regular,
+    // Inside block key context
+    BlockKey,
+    // Outside flow context
+    FlowOut,
+    // Inside flow context
+    FlowIn,
+    // Inside flow key context
+    FlowKey,
+}
+
+impl LexContext for YamlLexContext {
+    fn is_regular(&self) -> bool {
+        matches!(self, Self::Regular)
+    }
+}
+
+// https://yaml.org/spec/1.2.2/#rule-ns-plain-safe
+fn is_plain_safe(c: u8, context: YamlLexContext) -> bool {
+    use YamlLexContext::*;
+    match context {
+        FlowOut | BlockKey => is_non_space_char(c),
+        FlowIn | FlowKey => is_non_space_char(c) && !is_flow_indicator(c),
+        // Can happen when a YAML document starts with a plain token. This token will then be
+        // used by the parser to determine whether the parser is already inside the document
+        // body. This token is not used inside the CST and instead is just used to signal the
+        // start of document body
+        Regular => is_non_space_char(c) && !is_indicator(c),
+    }
+}
+
+// https://yaml.org/spec/1.2.2/#rule-ns-char
+fn is_non_space_char(c: u8) -> bool {
+    !is_space(c) && !is_break(c)
+}
+
+// https://yaml.org/spec/1.2.2/#rule-s-white
+fn is_space(c: u8) -> bool {
+    c == b' ' || c == b'\t'
+}
+
+// https://yaml.org/spec/1.2.2/#rule-b-char
+fn is_break(c: u8) -> bool {
+    c == b'\n' || c == b'\r'
+}
+
+fn is_indicator(c: u8) -> bool {
```

Comment:
```suggestion
// https://yaml.org/spec/1.2.2/#53-indicator-characters
fn is_indicator(c: u8) -> bool {
```

Would it be useful to add the YAML spec link here as well?